### ISSUE TITLE
Performance degrades when trained with custom dataset having rotated bounding boxes

### ISSUE THREAD

**Issue by saurabh-2905** [eriklindernoren/PyTorch-YOLOv3#611]
Hello @Flova ,

I am a Masters Student from TU Chemnitz, currently working on my master thesis in collaboration with the University of Bremen. My goal is to train an object detection model on synthetic images using UDA methods. For this, I am using this implementation of yolov3 as my base architecture. I have achieved quite a good mAP (0.85) when I trained the model on the custom dataset (real-life images) with normal bbox. The network is also able to perform detection relative to the achieved mAP.  

However, I am trying to use the network for fisheye images, so I have my dataset annotated with rotated bounding boxes. I have made the required changes that allow the network to handle the rotated bounding boxes. These changes are based on the paper 'Rotation Aware People Detection in Overhead Fisheye Images (Rapid)'. 

The changes majorly include,
1. Changing the config file to get extra parameter for the angle of rotation in output.
2. Changes in the dataloader class to get the new bbox (x, y, w, h, angle)
3. Additional loss function for prediction of angle.
4. New function `iou_rotated` to calculate the IOU of rotated bounding boxes, in place of `bbox_iou` 
5. Other minor changes to make network compatible for new bbox in `train.py`, `test.py` and other related files

The changes made can be seen in my repository: [https://gitlab.informatik.uni-bremen.de/digipath/band-saurabh/-/tree/rotated_new](url)  

I confirmed, that the changes made are working correctly by overfitting the model on 4-5 images. In this test, the network is able to predict the rotated bounding boxes with help of the `angle_loss` introduced. However, when I try to train the network on a set of images, the mAP goes down to 0.4 - 0.5.  I tried using various learning rates and changing other hyperparameters, but the max mAP achieved till now is around 0.5.

 Is there any specific part of the network that should be looked after, while making such changes in this case? Am I missing out on something?

Looking forward to your reply. Thank you.

**Comment by Flova** [comment]
Moin @saurabh-2905 ,

I haven't tried rotational bounding boxes yet. But I have the following suggestions:

1. I would try to deactivate the simple data augmentation and multiscale training.
2. Note that the hyperparams need to be changed in code not in the `.cfg` at the moment.
3. I heard that rotational YOLO works better if you have two outputs one for the `sin` and one for the `cos`. Sadly I don't find the source at the moment.
4. Try SGD instead of ADAM optimizer. It should take longer / more epochs but be a bit more stable.
5. Maybe there is some little bug somewhere in the operations you introduced, this could be tricky to find/detect.
6. Did you use transfer learning? Have a look at the loading of the weights. Adding more outputs could mess with the loading of the weights.

Nevertheless, a cool project, and I hope you find the issue soon. Let me know if you found it.

Good luck and cheers
Flo

**Comment by saurabh-2905** [comment]
Moin @Flova,

Thank you for the suggestion, I would try to find the problem by keeping these things in mind. 

Yes, I am not able to use pre-trained weights for the whole network due to the changes in the config file, but I am still able to use the weights for the backbone network (Darknet).

> 3\. I heard that rotational YOLO works better if you have two outputs one for the `sin` and one for the `cos`. Sadly I don't find the source at the moment.

Also, could you please elaborate on this point, if you find some source? It would really be of great help if you could.

Thanks for the reply. Will keep you updated about it.

Regards,
Saurabh

**Comment by Flova** [comment]
This discussion sums up the idea https://stats.stackexchange.com/questions/218407/encoding-angle-data-for-neural-network. It is not the YOLO specific one I was looking for but it's close enough I guess. 
Did you also adapt the nms?

Flo

**Comment by saurabh-2905** [comment]
Thanks for the link. I would go through it. I am sure it can be of some help to me.

Regarding the nms, yes I made some changes that I thought were required for this implementation. The code snippet of nms looks like this:
```python 
def non_max_suppression(prediction, conf_thres=0.5, nms_thres=0.4):
    """
    Removes detections with lower object confidence score than 'conf_thres' and performs
    Non-Maximum Suppression to further filter detections.
    Returns detections with shape:
        (x1, y1, x2, y2, angle, object_conf, class_score, class_pred)
    """

    # From (center x, center y, width, height) to (x1, y1, x2, y2)
    prediction[..., :4] = xywh2xyxy(prediction[..., :4])
    output = [None for _ in range(len(prediction))]
    for image_i, image_pred in enumerate(prediction):
        # Filter out confidence scores below threshold
        image_pred = image_pred[image_pred[:, 5] >= conf_thres]
        # If none are remaining => process next image
        if not image_pred.size(0):
            continue
        # Object confidence times class confidence
        score = image_pred[:, 5] * image_pred[:, 6:].max(1)[0]
        # Sort by it
        image_pred = image_pred[(-score).argsort()]
        class_confs, class_preds = image_pred[:, 6:].max(1, keepdim=True)
        detections = torch.cat((image_pred[:, :6], class_confs.float(), class_preds.float()), 1)
        # Perform non-maximum suppression
        keep_boxes = []
        while detections.size(0):
            large_overlap = iou_rotated(detections[0, :5].unsqueeze(0), detections[:, :5]) > nms_thres
            label_match = detections[0, -1] == detections[:, -1]
            # Indices of boxes with lower confidence scores, large IOUs and matching labels
            invalid = large_overlap & label_match
            weights = detections[invalid, 5:6]
            # Merge overlapping bboxes by order of confidence
            detections[0, :4] = (weights * detections[invalid, :4]).sum(0) / weights.sum()
            keep_boxes += [detections[0]]
            detections = detections[~invalid]
        if keep_boxes:
            output[image_i] = torch.stack(keep_boxes)
    
    # for o_i, out in enumerate(output):
    #     if out == None:
    #         output[o_i] = torch.zeros(1,8)
            

    return output
 ```
In the above code, the input variable `prediction` has values `(x1, y1, x2, y2, angle, conf, classes)`. The indexing of the variables in the function is changed according to the input parameters to give the appropriate values everywhere. I also wrote a new function to calculate `large_overlap` which takes into consideration the angle of rotation and calculates new co-ordinates of bbox after rotation to calculate the IOU. 
However, I was not sure if there are any other changes required? Especially, in the line `detections[0, :4] = (weights * detections[invalid, :4]).sum(0) / weights.sum()`.

Please let me know if this is correct or if there are any further changes required.

Cheers,
Saurabh

**Comment by Flova** [comment]
Sorry I currently don't have the time to investigate issues not directly related to this project. I, therefore, close this issue now. Good luck with your implementation in the future! If anyone has additional points/comments feel free to reopen.

Cheers,
Flo
